cloudsync_onedrive.py

```diff
@@ -872,35 +877,51 @@ def create(self, path, file_like, metadata=None) -> 'OInfo':
             r = self._upload_large(api_path, file_like, conflict="fail")
             return self._info_from_rest(r, root=self.dirname(path))
 
-    def _upload_large(self, drive_path, file_like, conflict):  # pylint: disable=too-many-locals
+    def _create_upload_session(self, api_path: str, conflict: str) -> str:
+        """Returns the url for a new upload session"""
+        resp = self._direct_api("post", f"{api_path}/createUploadSession",
+                                json={"item": {"@microsoft.graph.conflictBehavior": conflict}})
+        return resp["uploadUrl"]
+
+    def _upload_large(self, api_path, file_like, conflict):  # pylint: disable=too-many-locals
         size = _get_size_and_seek0(file_like)
         with self._api():
-            r = self._direct_api("post", f"{drive_path}/createUploadSession", json={"item": {"@microsoft.graph.conflictBehavior": conflict}})
-            upload_url = r["uploadUrl"]
-
+            upload_url = self._create_upload_session(api_path, conflict)
             data = file_like.read(self.upload_block_size)
-
+            max_retries = 3
             max_retries_per_block = 10
-
             cbfrom = 0
             retries = 0
+            retries_per_block = 0
             while data:
                 clen = len(data)             # fragment content size
                 cbto = cbfrom + clen - 1     # inclusive content byte range
                 cbrange = f"bytes {cbfrom}-{cbto}/{size}"
                 try:
                     headers = {"Content-Length": clen, "Content-Range": cbrange}
                     r = self._direct_api("put", url=upload_url, data=data, headers=headers)
-                except (CloudDisconnectedError, CloudTemporaryError) as e:
+                except OneDriveResourceModifiedError:
+                    # this error indicates that our upload session has been invalidated - restart the upload
```

Comment:
I'm not confident that this is the appropriate course of action. Based on my understanding of what's going on here, it's excellent to create a new error for this condition, but my instinct is to handle the error in the client. If the client is CloudSync, treat this new error as a temporary error. 

In CloudSync, temporary errors also result in retries, but they happen in a manner that is aware of the contract offered by the whole synchronization process. If this file has changed while uploading, is it appropriate to simply overwrite the changed file with whatever is being uploading? That looks to me like it should be treated as a sync conflict. If we retry the sync with the ResourceModifiedError, by treating that as a temporary error in CloudSync, it would enable CloudSync to resolve the conflict by renaming the conflicted file, or merging the changes, or whatever, when the retry occurs. If we simply overwrite the changed file, then the user that saved the changes to that file will find that, not only do their changes not exist in OneDrive, if they are also syncing this folder, that their changes have been lost locally as well.

Reply:
made the change.